### PR TITLE
Implemented access token authentication

### PR DESCRIPTION
**Pull request by Eugineer0**

### CODE REVIEW

File: Backend/CarRentalApp/CarRentalApp/Models/DTOs/Requests/UserLoginDTO.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+        [Required]
+        [RegularExpression(
+            @"^" +
+            "(?=.*[A-Z])" + // At least one UpperCase
+            "(?=.*[0-9])" + // At least one Digit
+            "(?=.*[a-z])" + // At least one LowerCase
+            ".{5,25}" +     // Length between 5 and 25
+            "$",
+         ErrorMessage = "Incorrect format")]
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Models/DTOs/Requests/UserLoginDTO.cs:18]: How would user know what exactly is incorrect? Also why checking this when you are trying to login, i.e. password already set for this user?

File: Backend/CarRentalApp/CarRentalApp/Services/Data/UserRepositoryService.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        public async Task<User?> GetByEmailAsync(string email)
+        {
+            throw new NotImplementedException();
+        }
+
+        public async Task<User?> GetByUsernameAsync(string username)
+        {
+            throw new NotImplementedException();
+        }
+    }
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Data/UserRepositoryService.cs:22]: How does it work if all methods throw NotImplementedException?

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    public class UserService : IUserService
+    {
+        private IUserRepository _userRepository;
+
+        public UserService(IUserRepository userRepository)
+        {
+            _userRepository = userRepository;
+        }
+
+        public async Task<User?> GetValidUserAsync(UserLoginDTO user2Login)
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs:18]: What is the definition of "valid user"? Is it the responsibility of user service to check user passwords?

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            var hashedPassword = DigestPassword(userDTO.Password, existingUser.Salt);
+            return hashedPassword.SequenceEqual(existingUser.HashedPassword);
+        }
+
+        private byte[] DigestPassword(string password, string salt)
+        {
+            byte[] data = Encoding.UTF8.GetBytes(password + salt);
+
+            HashAlgorithm sha = SHA256.Create();
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs:39]: Why SHA256? Is it secure enough? have you considered any alternatives?

File: Backend/CarRentalApp/CarRentalApp/Services/Token/ITokenService.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+﻿using System.IdentityModel.Tokens.Jwt;
+using CarRentalApp.Models.Data;
+
+namespace CarRentalApp.Services.Token
+{
+    public interface ITokenService
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Token/ITokenService.cs:6]: Why do you need this interface?

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/IUserService.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿using CarRentalApp.Models.DTOs.Requests;
+using CarRentalApp.Models.Data;
+
+namespace CarRentalApp.Services.Identity
+{
+    public interface IUserService
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Identity/IUserService.cs:6]: Why do you need this interface?

File: Backend/CarRentalApp/CarRentalApp/Services/Authentication/IAuthenticationService.cs [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿using CarRentalApp.Models.DTOs.Requests;
+using CarRentalApp.Models.DTOs.Responces;
+
+namespace CarRentalApp.Services.Authentication
+{
+    public interface IAuthenticationService
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Authentication/IAuthenticationService.cs:6]: Why do you need this interface?

File: Backend/CarRentalApp/CarRentalApp/Services/Token/TokenService.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        public TokenService(IConfiguration configuration)
+        {
+            _configuration = configuration;
+        }
+
+        public JwtSecurityToken GenerateAccessToken(User user)
+        {
+            var accessJwtGenerationParams = new TokenGenerationParameters();
+
+            _configuration.GetSection("AccessJwtConfig:GenerationParameters").Bind(accessJwtGenerationParams);
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Token/TokenService.cs:23]: Use Options pattern please

File: Backend/CarRentalApp/CarRentalApp/Services/Data/IUserRepository.cs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿using CarRentalApp.Models.Data;
+
+namespace CarRentalApp.Services.Data
+{
+    public interface IUserRepository
```
Comment (andrei-kruhliakou-itechart) [inline Backend/CarRentalApp/CarRentalApp/Services/Data/IUserRepository.cs:5]: What do you need repository for?

File: Backend/CarRentalApp/CarRentalApp/Services/Token/ITokenService.cs (hunk 5 above)
Comment (Eugineer0) [reply Backend/CarRentalApp/CarRentalApp/Services/Token/ITokenService.cs:6]: This interface is required by DI pattern

File: Backend/CarRentalApp/CarRentalApp/Services/Data/IUserRepository.cs (hunk 9 above)
Comment (Eugineer0) [reply Backend/CarRentalApp/CarRentalApp/Services/Data/IUserRepository.cs:5]: I am using this repository as layer between UserService and DbContext

File: Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs (hunk 3 above)
Comment (Eugineer0) [reply Backend/CarRentalApp/CarRentalApp/Services/Identity/UserService.cs:18]: > Is it the responsibility of user service to check user passwords? 

I decided to unite all operations with users in this service

**Pull request closed without merging**